idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirReferenceShortener.kt

```diff
@@ -303,52 +365,48 @@ private class ElementsToShortenCollector(
-            val firstFoundClass = shorteningContext.findFirstClassifierInScopesByName(positionScopes, classId.shortClassName)?.classId
-
-            if (firstFoundClass == classId) {
-                return ShortenQualifier(qualifier)
-            }
-        }
+            val option = classShortenOption(classId.toRegularClass() ?: return null)
+            if (option == ShortenOption.DO_NOT_SHORTEN) continue
 
-        val (mostTopLevelClassId, mostTopLevelQualifier) = allClassIds.zip(allQualifiers).last()
-        val availableClassifier = shorteningContext.findFirstClassifierInScopesByName(positionScopes, mostTopLevelClassId.shortClassName)
+            // Find class with the same name that's already available in this file.
+            val availableClassifier = shorteningContext.findFirstClassifierInScopesByName(positionScopes, classId.shortClassName)
 
-        check(availableClassifier?.classId != mostTopLevelClassId) {
-            "If this condition were true, we would have exited from the loop on the last iteration. ClassId = $mostTopLevelClassId"
-        }
-
-        return if (availableClassifier == null || availableClassifier.isFromStarOrPackageImport) {
-            ShortenQualifier(mostTopLevelQualifier, mostTopLevelClassId.asSingleFqName())
-        } else {
-            findFakePackageToShorten(mostTopLevelQualifier)
+            when {
+                availableClassifier == null -> {
+                    if (option == ShortenOption.SHORTEN_IF_ALREADY_IMPORTED) continue
+                    return ShortenQualifier(qualifier, classId.asSingleFqName(), option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT)
+                }
+                availableClassifier.symbol == classId -> {
+                    return when {
+                        !availableClassifier.isFromPackageImport && !availableClassifier.isFromStarImport &&
+                                option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT -> {
+                            ShortenQualifier(qualifier, classId.asSingleFqName(), true)
+                        }
+                        else -> ShortenQualifier(qualifier)
+                    }
+                }
+                // Allow using star import to overwrite members implicitly imported by default.
+                availableClassifier.isFromDefaultImport && availableClassifier.isFromStarImport && option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT -> {
+                    return ShortenQualifier(qualifier, classId.asSingleFqName(), true)
+                }
+                // Allow using explicit import to overwrite members imported on-demand or in package
+                (availableClassifier.isFromStarImport || availableClassifier.isFromPackageImport) && option == ShortenOption.SHORTEN_AND_IMPORT -> {
+                    return ShortenQualifier(qualifier, classId.asSingleFqName(), false)
+                }
+            }
```

Comment:
All these additions look quite similar to other additions in `findTypeToShorten`, except for the creation of `ShortenType` v.s. `ShortenQualifier`? Won't it be better to utilize/unify these to guarantee the same behavior, less code, etc.?

Reply:
Sounds good! Let me unify these.